_posts/2023-06-15-our-journey-at-f5-with-apache-arrow-part-2.md

```diff
@@ -0,0 +1,199 @@
+
+```go
+var (
+  // Simplified schema definition generated by the Arrow Record encoder based on 
+  // the data observed.
+  TracesSchema = arrow.NewSchema([]arrow.Field{
+    {Name: constants.ID, Type: arrow.PrimitiveTypes.Uint16, Nullable: true},
+    {Name: constants.StartTimeUnixNano, Type: arrow.FixedWidthTypes.Timestamp_ns},
+    {Name: constants.TraceId, Type: &arrow.FixedSizeBinaryType{ByteWidth: 16}},
+    {Name: constants.SpanId, Type: &arrow.FixedSizeBinaryType{ByteWidth: 8}},
+    {Name: constants.Name, Type: &arrow.DictionaryType {
+      IndexType: arrow.PrimitiveTypes.Uint8,
+      ValueType: arrow.BinaryTypes.String}},
+    {Name: constants.KIND, Type: &arrow.DictionaryType {
+      IndexType: arrow.PrimitiveTypes.Uint8,
+      ValueType: arrow.PrimitiveTypes.Int32,
+    }, Nullable: true},
+  }, nil)
+)
+```
+
+Additionally, we desire a system capable of automatically adapting the aforementioned schema if it encounters new fields or existing fields with a cardinality exceeding the size of the current dictionary definition in future batches. In extreme scenarios, if the cardinality of a specific field surpasses a certain threshold, we would prefer the system to automatically revert to the non-dictionary representation (mechanism of dictionary overflow). That is precisely what we will elaborate on in the remainder of this section.  
+
+An overview of the different components and events used to implement this approach is depicted in figure 1.
+
+<figure style="text-align: center;">
+  <img src="{{ site.baseurl }}/img/journey-apache-arrow/adaptive-schema-architecture.svg" width="100%" class="img-responsive" alt="Fig 1: Adaptive Arrow schema architecture overview.">
+  <figcaption>Fig 1: Adaptive Arrow schema architecture overview.</figcaption>
+</figure>
+
+The overall Adaptive Arrow schema component takes a data stream segmented into batches and produces one or multiple streams of Arrow Records (one schema per stream). Each of these records is defined with an Arrow schema, which is based both on the annotated Arrow schema and the shape of fields observed in the incoming data.
+
+More specifically, the process of the Adaptive Arrow schema component consists of four main phases
+
+**Initialization phase**
+
+During the initialization phase, the Arrow Record Encoder reads the annotated Arrow schema (i.e. the reference schema) and generates a collection of transformations. When these transformations are applied to the reference schema, they yield the first minimal Arrow schema that adheres to the constraints depicted by these annotations. In this initial iteration, all optional fields are eliminated, and all dictionary-encoded fields are configured to utilize the smallest encoding as defined by the annotation. These transformations form a tree, reflecting the structure of the reference schema. 
```

Comment:
> and all dictionary-encoded fields are configured to utilize the smallest encoding as defined by the annotation.

I was a little confused about this wording as the example above *only*  shows dictionaries using Dictionary8 (rather than some combination of smaller and larger encodings)

Reply:
Made a change to clarify a bit.